rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java

```diff
@@ -1175,7 +1067,285 @@ private static String extractCaptureGroupName(char[] src, int cpbegin, int cpend
+            /* UnicodeEscapeSequence */
+            case 'u':
+                nDigits += 2;
+            /* fall through */ case 'x': /* HexEscapeSequence */
+                {
+                    int n = 0;
+                    int i;
+                    for (i = 0; (i < nDigits) && (state.cp < state.cpend); i++) {
+                        c = src[state.cp++];
+                        n = Kit.xDigitToInt(c, n);
+                        if (n < 0) {
+                            // Back off to accepting the original
+                            // 'u' or 'x' as a literal
+                            state.cp -= (i + 2);
+                            n = src[state.cp++];
+                            break;
+                        }
+                    }
+                    c = (char) n;
+                }
+                doFlat(state, c);
+                break;
+            /* Character class escapes */
+            case 'd':
+                state.result = new RENode(REOP_DIGIT);
+                state.progLength++;
+                break;
+            case 'D':
+                state.result = new RENode(REOP_NONDIGIT);
+                state.progLength++;
+                break;
+            case 's':
+                state.result = new RENode(REOP_SPACE);
+                state.progLength++;
+                break;
+            case 'S':
+                state.result = new RENode(REOP_NONSPACE);
+                state.progLength++;
+                break;
+            case 'w':
+                state.result = new RENode(REOP_ALNUM);
+                state.progLength++;
+                break;
+            case 'W':
+                state.result = new RENode(REOP_NONALNUM);
+                state.progLength++;
+                break;
+            /* IdentityEscape */
+            default:
+                state.cp--;
+                return parseIdentityEscape(state, params);
+        }
+
+        return true;
+    }
+
+    // to be called when the current and next characters are '0'
+    private static void parseMultipleLeadingZerosAsOctalEscape(CompilerState state) {
+        char[] src = state.cpbegin;
+        int num = 0;
+        char c;
+        reportWarning(state.cx, "msg.bad.backref", "");
+        while (num < 040 && state.cp < state.cpend) {
+            c = src[state.cp];
+            if ((c >= '0') && (c <= '7')) {
+                state.cp++;
+                num = 8 * num + (c - '0');
+            } else break;
+        }
+        c = (char) num;
+        doFlat(state, c);
+    }
+
+    static class ClassContents {
+        boolean sense;
+        ArrayList<Character> chars;
+        ArrayList<Character> ranges; // ranges stored as (start1, end1, start2, end2, ...)
+        ArrayList<RENode> escapeNodes;
+
+        ClassContents() {
+            chars = new ArrayList<>();
+            ranges = new ArrayList<>();
+            escapeNodes = new ArrayList<>();
+            sense = true;
+        }
+    }
```

Comment:
Why not do this in the field declaration?

```suggestion
    static class ClassContents {
        boolean sense = true;
        ArrayList<Character> chars = new ArrayList<>();
        ArrayList<Character> ranges = new ArrayList<(); // ranges stored as (start1, end1, start2, end2, ...)
        ArrayList<RENode> escapeNodes = new ArrayList<>();
    }
```